### PR TITLE
dialects: Add support for stencil inlining

### PR DESCRIPTION
**Pull request by meshtag**
This PR intends to add the first draft of stencil inlining support. 

Screenshot from 2023-04-20 18-01-20

Screenshot from 2023-04-20 18-43-31

### CODE REVIEW

**Comment by PapyChacal** [comment]
Looks promising! Do you think you could factor a few things out of the main inlining function?

File: xdsl/transforms/experimental/StencilInlining.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+            for use in res.uses:
+                if (isinstance(use.operation, ApplyOp)):
+                    return use.operation
+        return None
+
+
+@dataclass
+class InliningRewrite(StencilInliningPattern):
+
+    def InlineProducer(self, producer_op: ApplyOp, rewriter: PatternRewriter,
```
Comment (AntonLydike) [inline xdsl/transforms/experimental/StencilInlining.py:65]: This is quite a lot, could you try and split this into multiple smaller functions that have descriptive names and docstrings? I think that would make it easier to see what's going on and in the end review it.

File: xdsl/transforms/experimental/StencilInlining.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    def HasSingleConsumer(self, producer_op: ApplyOp) -> bool:
+        applyop_consumers_num = 0
+        for res in producer_op.res:
+            for use in res.uses:
+                if (isinstance(use.operation, ApplyOp)):
+                    applyop_consumers_num += 1
+                if (applyop_consumers_num > 1):
+                    break
+
+        return applyop_consumers_num == 1
```
Comment (superlopuh) [inline xdsl/transforms/experimental/StencilInlining.py:30]:
```suggestion
        return any(isinstance(use.operation, ApplyOp) for res in producer_op.res for use in res.uses)
```

File: tests/xdsl_opt/test_xdsl_opt.py [hunk 3]
```diff
@@ -14,12 +14,9 @@ def test_opt():
...
     assert list(opt.available_passes.keys()) == [
-        'lower-mpi',
-        'convert-stencil-to-ll-mlir',
-        'convert-stencil-to-gpu',
-        'stencil-shape-inference',
-        'stencil-to-local-2d-horizontal',
-        'frontend-desymrefy',
+        'lower-mpi', 'convert-stencil-to-ll-mlir', 'convert-stencil-to-gpu',
+        'stencil-shape-inference', 'stencil-to-local-2d-horizontal',
+        'frontend-desymrefy', 'stencil-inlining'
```
Comment (superlopuh) [inline tests/xdsl_opt/test_xdsl_opt.py:19]:
```suggestion
        'frontend-desymrefy', 'stencil-inlining',
```

File: xdsl/transforms/experimental/StencilInlining.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        consumer_op_parent = consumer_op.parent
+        assert isinstance(consumer_op_parent, Block)
+        consumer_op_parent.erase_op(consumer_op, False)
+
+        # Remove producer op from the IR.
+        rewriter.erase_matched_op(False)
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: ApplyOp, rewriter: PatternRewriter, /):
+        if super().HasSingleConsumer(op) and super().IsStencilInliningPossible(
```
[the remark this answers is not in the document]
Comment (meshtag) [reply xdsl/transforms/experimental/StencilInlining.py]: Usually, I prefer making it explicitly known that the called method is from the parent class, are there any downsides to it? 
I don't have a strong opinion on this and will change it if you think that makes things better.

File: xdsl/transforms/experimental/StencilInlining.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+                if not len(inlined_op_return_arguments):
+                    op_clone = op.clone()
+                    inlined_op_block.add_op(op_clone)
+                else:
+                    combined_list: list[SSAValue] = [
+                        *inlined_op_return_arguments, *list(op.operands)
+                    ]
+
+                    inlined_op_return = ReturnOp.get(combined_list)
+                    inlined_op_block.add_op(inlined_op_return)
```
Comment (PapyChacal) [inline xdsl/transforms/experimental/StencilInlining.py:290]: This specific part is still quite hard to get into from the outside. Could you maybe add high level comments as a first step?

**Comment by tobiasgrosser** [comment]
@meshtag and @PapyChacal, what is the status for this PR?

**Comment by meshtag** [comment]
@PapyChacal proposed a different method to do this, so we'll review that one once it is ready. This PR might serve as a source for filecheck tests or something baseline to compare against.

**Comment by tobiasgrosser** [comment]
This PR seems old and outdated. I close it for now. Please reopen if that ever becomes relevant again.

**Pull request closed without merging**